src/simulation/dynamics/_GeneralModuleFiles/svIntegratorRungeKutta.h

```diff
@@ -0,0 +1,208 @@
+     * needed, where each "k" coefficient has the same size as the state.
+     * This type allows us to store these "k" coefficients.
+     */
+    using KCoefficientsValues = std::array<ExtendedStateVector, numberStages>;
+
+    /**
+     * Computes the derivatives of every state given a time and current states.
+     *
+     * Internally, this sets the states on the dynamic objects and
+     * calls the equationsOfMotion methods.
+     */
+    ExtendedStateVector computeDerivatives(double time, double timeStep,
+                                           const ExtendedStateVector& states);
+
+    /**
+     * Computes the "k" coefficients of the Runge-Kutta method
+     * for a time and state.
+     */
+    KCoefficientsValues computeKCoefficients(double currentTime, double timeStep,
+                                             const ExtendedStateVector& currentStates);
+
+    /**
+     * Adds the "k" coefficients, weighted by the "c" coefficients
+     * to find the state after the time step.
+     */
+    ExtendedStateVector computeNextState(double timeStep, const ExtendedStateVector& currentStates,
+                                         const KCoefficientsValues& kVectors);
+
+  protected:
+    // coefficients is stored as a pointer to support polymorphism
+    /** Coefficients to be used in the method */
+    const std::unique_ptr<RKCoefficients<numberStages>> coefficients;
+};
+
+template <size_t numberStages>
+svIntegratorRungeKutta<numberStages>::svIntegratorRungeKutta(
+    DynamicObject* dynIn, const RKCoefficients<numberStages>& coefficients)
+    : StateVecIntegrator(dynIn),
+      coefficients(std::make_unique<RKCoefficients<numberStages>>(coefficients))
+{
+}
+
+template <size_t numberStages>
+svIntegratorRungeKutta<numberStages>::svIntegratorRungeKutta(
+    DynamicObject* dynIn, std::unique_ptr<RKCoefficients<numberStages>>&& coefficients)
+    : StateVecIntegrator(dynIn), coefficients(std::move(coefficients))
+{
+}
+
+template <size_t numberStages>
+void svIntegratorRungeKutta<numberStages>::integrate(double currentTime, double timeStep)
+{
+    ExtendedStateVector currentState = ExtendedStateVector::fromStates(this->dynPtrs);
+    KCoefficientsValues kValues = this->computeKCoefficients(currentTime, timeStep, currentState);
+    ExtendedStateVector nextState = this->computeNextState(timeStep, currentState, kValues);
+    nextState.setStates(this->dynPtrs);
+}
+
+template <size_t numberStages>
+ExtendedStateVector
+svIntegratorRungeKutta<numberStages>::computeDerivatives(double time, double timeStep,
+                                                         const ExtendedStateVector& states)
+{
+    states.setStates(this->dynPtrs);
+
+    for (auto dynPtr : this->dynPtrs) {
+        dynPtr->equationsOfMotion(time, timeStep);
+    }
+
+    return ExtendedStateVector::fromStateDerivs(this->dynPtrs);
+}
+
+template <size_t numberStages>
+auto svIntegratorRungeKutta<numberStages>::computeKCoefficients(
```

Comment:
It's a trailing return type:

```
auto svIntegratorRungeKutta<numberStages>::computeKCoefficients(...) -> KCoefficientsValues
```

because `KCoefficientsValues` is only understood by the compiler once it has seen that the method is defined for `svIntegratorRungeKutta<numberStages>` , which defines `KCoefficientsValues`. An alternative signature would be:

```
svIntegratorRungeKutta<numberStages>::KCoefficientsValues svIntegratorRungeKutta<numberStages>::computeKCoefficients(...)
```
But I personally find that trailing return types create less clutter and it's easier to see the return type for complex signatures. I could change it if we want to stay clear from trailing return types.